Implement the instance picker

thibaultamartin opened the pull request:
This work is largely based on Austin Huang's great work on servers.joinmatrix.org

thibaultamartin commented:
## Data Model for instances

The two options I have in mind are:
1. A single toml file that any interested party can easily consume.
2. One .md file per instance, with structure information in the frontmatter.

### Single toml file approach

Pros/cons

### One markdown file per instance

Pros/cons

jplatte commented:
> More difficult to browse and parse for third parties

Zola is not limited to HTML as an output format. It should be possible to write some Tera code to expose the data stored in frontmatter of multiple files as one JSON file, for example.

thibaultamartin commented:
Is it possible?! That would entirely save the day.

If you have pointers to the right docs that would be very helpful

jplatte commented:
Uhh, now I'm wondering if I am imagining things 😅 
I know it used to work in *a* static site generator I used at some point, might have been Jekyll where every content file was processed as long as it contained frontmatter? Certainly not seeing anything in docs – it's not even documented that you can use `.html` files instead of `.md`? I'm pretty damn sure Zola supports that though.

thibaultamartin commented:
## Updating the CI data

The only bits that should be updated automatically are the software used and the version it relies on.
All of this information can be retrieved via the [version endpoint](https://spec.matrix.org/v1.8/server-server-api/#get_matrixfederationv1version).

Typically for ergaster.org:

```
thib@MacBook ~ % curl -s https://ergaster.org/.well-known/matrix/server | jq '."m.server"'
"matrix.ergaster.org:443"
thib@MacBook ~ % curl -s https://matrix.ergaster.org/_matrix/federation/v1/version | jq '.server.name'
"Synapse"
thib@MacBook ~ % curl -s https://matrix.ergaster.org/_matrix/federation/v1/version | jq '.server.version'
"1.91.1"
```

Then either the toml file needs to be parsed and updated, or the markdown files need to be parsed and updated depending on the data model chosen

thibaultamartin commented:
Current state of things. It's not responsive yet.

<img width="1440" alt="Screenshot 2023-09-07 at 16 51 44" src="https://github.com/matrix-org/matrix.org/assets/1688309/ee18de0e-f526-4fbc-970c-e09aa9039972">

jplatte commented on .github/workflows/update-public-instances.yml:
```diff
@@ -0,0 +1,48 @@
+name: Update public instances
+on:
+  push:
+    branches:
+      - instance-picker
```
Because of this, your CI run doesn't count as a PR run. You need to either change the `github.event_name` check below to `push`, or change this to

```suggestion
  pull_request:
    branches:
      - main
```

(triggered by  PR to the main branch)

jplatte commented on .github/workflows/update-public-instances.yml:
```diff
@@ -0,0 +1,48 @@
+          cp piu/instances.toml morg/content/public-instances/instances.toml
+
+      - name: Commit and push data (within PR)
+        if: ${{ github.event_name == 'pull_request' }}
```
No need for `%{{` inside `if`

```suggestion
        if: github.event_name == 'pull_request'
```

jplatte commented on .github/workflows/update-public-instances.yml:
```diff
@@ -0,0 +1,48 @@
+        run: |
+          git config --global user.name "GitHub CI"
+          git config --global user.email "noreply@matrix.org"
+          git switch ${ github.ref_name }
```
```suggestion
          git switch $GITHUB_REF_NAME
```

or:

```suggestion
          git switch ${{ github.ref_name }}
```

jplatte commented on .github/workflows/update-public-instances.yml:
```diff
@@ -0,0 +1,48 @@
+        working-directory: morg
+
+      - name: Commit and push data (scheduled)
+        if: ${{ github.event_name == 'schedule' }}
```
```suggestion
        if: github.event_name == 'schedule'
```

jplatte commented on .github/_typos.toml:
```diff
@@ -3,4 +3,5 @@ extend-exclude = [
 	"*.js*", "*.json", "*.svg",
 	"/static/jira",
     "/content/blog/2023/07/2023-07-28-twim.md"
+    "/content/public-instances/instances.toml"
```
```suggestion
    "/content/blog/2023/07/2023-07-28-twim.md",
    "/content/public-instances/instances.toml",
```
